exchanges/stream/websocket_test.go

```diff
@@ -0,0 +1,956 @@
+	ws := *New()
+	err := ws.Setup(defaultSetup)
+	if err != nil {
+		t.Fatal(err)
+	}
+	ws.trafficTimeout = time.Minute
+	ws.connector = func() error { return nil }
+
+	err = ws.Connect()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	ws.TrafficAlert <- struct{}{}
+
+	timer := time.NewTimer(900 * time.Millisecond)
+	ws.ReadMessageErrors <- errors.New("errorText")
+	select {
+	case err := <-ws.ToRoutine:
+		if err.(error).Error() != "errorText" {
+			t.Errorf("Expected 'errorText', received %v", err)
+		}
+	case <-timer.C:
+		t.Error("Timeout waiting for datahandler to receive error")
+	}
+	ws.ReadMessageErrors <- &websocket.CloseError{
+		Code: 1006,
+		Text: "errorText",
+	}
+outer:
+	for {
+		select {
+		case <-ws.ToRoutine:
+			t.Fatal("Error is a disconnection error")
+		case <-timer.C:
+			break outer
+		}
+	}
+}
+
+func TestWebsocket(t *testing.T) {
+	ws := Websocket{}
+	err := ws.Setup(&WebsocketSetup{
+		ExchangeName: "test",
+		Enabled:      true,
+	})
+	if err != nil && err.Error() != "test Websocket already initialised" {
+		t.Errorf("Expected 'test Websocket already initialised', received %v", err)
+	}
+
+	ws = *New()
+	err = ws.SetProxyAddress("garbagio")
+	if err == nil {
+		t.Error("error cannot be nil")
+	}
+	err = ws.SetProxyAddress("https://192.168.0.1:1337")
+	if err != nil {
+		t.Error("SetProxyAddress", err)
+	}
+	// removing proxy
+	err = ws.SetProxyAddress("")
+	if err != nil {
+		t.Error(err)
+	}
+	// reinstate proxy
+	err = ws.SetProxyAddress("http://localhost:1337")
+	if err != nil {
+		t.Error(err)
+	}
+	// conflict proxy
+	err = ws.SetProxyAddress("http://localhost:1337")
+	if err == nil {
+		t.Error("error cannot be nil")
+	}
+	err = ws.Setup(defaultSetup)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if ws.GetName() != "exchangeName" {
+		t.Error("WebsocketSetup")
+	}
+
+	if !ws.IsEnabled() {
+		t.Error("WebsocketSetup")
+	}
+
+	ws.setEnabled(false)
+	if ws.IsEnabled() {
+		t.Error("WebsocketSetup")
+	}
+	ws.setEnabled(true)
+	if !ws.IsEnabled() {
+		t.Error("WebsocketSetup")
+	}
+
+	if ws.GetProxyAddress() != "http://localhost:1337" {
+		t.Error("WebsocketSetup")
+	}
+
+	if ws.GetWebsocketURL() != "wss://testRunningURL" {
+		t.Error("WebsocketSetup")
+	}
+	if ws.trafficTimeout != time.Second*5 {
+		t.Error("WebsocketSetup")
+	}
+	// -- Not connected shutdown
+	err = ws.Shutdown()
+	if err == nil {
+		t.Fatal("should not be connected to able to shut down")
+	}
+	// -- Normal connect
+	err = ws.Connect()
+	if err != nil {
+		t.Fatal("WebsocketSetup", err)
+	}
+	err = ws.SetWebsocketURL("ws://demos.kaazing.com/echo", false, false)
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = ws.SetWebsocketURL("ws://demos.kaazing.com/echo", true, false)
+	if err != nil {
+		t.Fatal(err)
+	}
+	// -- Already connected connect
+	err = ws.Connect()
+	if err == nil {
+		t.Fatal("should not connect, already connected")
+	}
+	// -- Normal shutdown
+	err = ws.Shutdown()
+	if err != nil {
+		t.Fatal("WebsocketSetup", err)
+	}
```

Comment:
Rather than remove the wg.wait, I think you could expand this test by adding it back again.
Once wg.wait has finished, you can then verify that connected status and connecting status is false